Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,598 @@
+                }.map {
+                    let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                    connection.isLeased = true
+                    return connection
+                }
+            }.and(handshakePromise.futureResult).map { (arg) -> ConnectionPool.Connection in
+                let (connection, _) = arg
+                self.configureCloseCallback(of: connection)
+                return connection
+            }.flatMapError { error in
+                let action = self.stateLock.withLock {
+                    self.state.failedConnectionAction()
+                }
+                switch action {
+                case .makeConnectionAndComplete(let el, let promise):
+                    self.makeConnection(on: el).cascade(to: promise)
+                case .none:
+                    break
+                }
+                return self.eventLoop.makeFailedFuture(error)
+            }
+        }
+
+        /// Adds a callback on connection close that asks the `state` what to do about this
+        ///
+        /// The callback informs the state about the event, and the state returns a
+        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                guard connection.shouldRunCloseCallback else {
+                    return
+                }
+                switch result {
+                case .success:
+                    let action = self.parentPool.connectionProvidersLock.withLock {
+                        self.stateLock.withLock { self.state.removeClosedConnection(connection) }
+                    }
+                    switch action {
+                    case .none:
+                        break
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func prepareForClose() {
+            let (waitersFutures, closeFutures) = self.stateLock.withLock { () -> ([EventLoopFuture<Connection>], [EventLoopFuture<Void>]) in
+                assert(self.state.activity == .opened, "Invalid activity: \(self.state.activity)")
+                self.state.activity = .closing
+                // Fail waiters
+                let waitersCopy = self.state.waiters
+                self.state.waiters.removeAll()
+                let waitersPromises = waitersCopy.map { $0.promise }
+                let waitersFutures = waitersPromises.map { $0.futureResult }
+                waitersPromises.forEach { $0.fail(HTTPClientError.cancelled) }
```

Comment:
@adtrevor this is a problem here. We're holding the lock and whilst holding the lock, we're calling out to users (by failing the promise). If the user calls into the connection pool from this, it'll deadlock/crash

Reply:
@adtrevor okay, in this very case it's actually okay because `prepareForClose` can't be called on any event loop so the `.fail(...)` won't be run under the lock. That's a bit subtle but okay.